param.go

```diff
@@ -0,0 +1,187 @@
+	param interface {
+		param()
+
+		// Human-friendly name for this param type. We use this exclusively in
+		// error messages.
+		ParamTypeName() string
+
+		// Comprehensive list of dependencies this parameter represents.
+		Dependencies() []edge
+	}
+
+	// paramList holds all arguments of the constructor as params.
+	paramList struct {
+		ctype reflect.Type // type of the constructor
+
+		Params []param
+	}
+
+	// paramSimple is an explicitly requested type, optionally with a name.
+	//
+	// This object must be present in the graph as-is unless it's specified as
+	// optional.
+	paramSimple struct {
+		Name     string
+		Optional bool
+		Type     reflect.Type
+	}
+
+	// paramObject is a dig.In struct where each field is another param.
+	//
+	// This object is not expected in the graph as-is.
+	paramObject struct {
+		Type   reflect.Type
+		Fields []paramObjectField
+
+		deps []edge
+	}
+
+	// paramObjectField is a single field of a dig.In struct.
+	paramObjectField struct {
+		// Name of the field specified via a name:".." tag.
+		Name string
+
+		// Index of this field in the target struct.
+		//
+		// We need to track this separately because not all fields of the
+		// struct map to params.
+		Index int
+
+		// The dependency requested by this field.
+		Param param
+	}
+)
+
+var (
+	_ param = paramSimple{}
+	_ param = paramObject{}
+	_ param = paramList{}
+)
+
+func (paramList) param()   {}
+func (paramSimple) param() {}
+func (paramObject) param() {}
+
+func (paramList) ParamTypeName() string   { return "parameter list" }
+func (paramSimple) ParamTypeName() string { return "simple type" }
+func (paramObject) ParamTypeName() string { return "parameter object" }
+
+// newParamList builds a paramList from the provided constructor type.
+//
+// Variadic arguments of a constructor are ignored and not included as
+// dependencies.
+func newParamList(ctype reflect.Type) (paramList, error) {
+	numArgs := ctype.NumIn()
+	if ctype.IsVariadic() {
+		// NOTE: If the function is variadic, we skip the last argument
+		// because we're not filling variadic arguments yet. See #120.
+		numArgs--
+	}
+
+	pl := paramList{
+		ctype:  ctype,
+		Params: make([]param, 0, numArgs),
+	}
+
+	for i := 0; i < numArgs; i++ {
+		p, err := newParam(ctype.In(i))
+		if err != nil {
+			return pl, errWrapf(err, "bad argument %d: %v", i+1)
+		}
+		pl.Params = append(pl.Params, p)
+	}
+	return pl, nil
+}
+
+// newParam builds a param from the given type. If the provided type is a
+// dig.In struct, an paramObject will be returned.
+func newParam(t reflect.Type) (param, error) {
+	if IsIn(t) {
+		return newParamObject(t)
+	}
+	return paramSimple{Type: t}, nil
+}
+
+// newParamObject builds an paramObject from the provided type. The type MUST
+// be a dig.In struct.
+func newParamObject(t reflect.Type) (paramObject, error) {
+	op := paramObject{Type: t}
+
+	for i := 0; i < t.NumField(); i++ {
+		f := t.Field(i)
+		if f.Type == _inType {
+			// Skip over the dig.In embed.
```

Comment:
uhhhhh, doesn't this break things?

Reply:
This doesn't break nested dig.Ins. (Test added.)

This is skipping the embed of the actual `dig.In` type.